http-body-util/src/throttle.rs

```diff
@@ -0,0 +1,163 @@
+use bytes::Buf;
+use http::HeaderMap;
+use http_body::{Body, SizeHint};
+use pin_project_lite::pin_project;
+use std::{
+    convert::{TryFrom, TryInto},
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    time::Duration,
+};
+use tokio::time::{sleep, Instant, Sleep};
+
+#[derive(Debug)]
+enum State {
+    Waiting(Pin<Box<Sleep>>, Instant),
+    Ready(Instant),
+    Init,
+}
+
+pin_project! {
+    /// A throttled body.
+    #[derive(Debug)]
+    #[cfg_attr(docsrs, doc(cfg(feature = "tokio")))]
+    pub struct Throttle<B> {
+        #[pin]
+        inner: B,
+        state: State,
+        cursor: f64,
+        byte_rate: f64,
+    }
+}
+
+impl<B> Throttle<B> {
+    /// Create a new `Throttle`.
+    ///
+    /// # Panic
+    ///
+    /// Will panic if milliseconds in `duration` is larger than `u32::MAX`.
+    pub fn new(body: B, duration: Duration, bytes: u32) -> Self {
+        let bytes = f64::from(bytes);
+        let duration = f64::from(u32::try_from(duration.as_millis()).expect("duration too large"));
```

Comment:
This could instead use `.as_secs_f64()`. This is a units change, but by using this function below, you can keep things aligned.

Reply:
Makes sense.